File: src/test/java/pl/training/module7/ReturnAsapEquivalenceTest.java
package pl.training.module7;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertSame;
import static org.junit.jupiter.api.Assertions.assertThrows;

import java.util.AbstractList;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import org.junit.jupiter.api.Test;

import pl.training.module7.returnasap.Artifact;
import pl.training.module7.returnasap.after.ArtifactFinder;
import pl.training.module7.returnasap.before.LegacyArtifactFinder;

final class ReturnAsapEquivalenceTest {
    private final LegacyArtifactFinder before = new LegacyArtifactFinder();
    private final ArtifactFinder after = new ArtifactFinder();

    @Test
    void earlyReturnPreservesEmptyMissingAndPresentResults() {
        var first = new Artifact("api.jar", "sha-1");
        var second = new Artifact("worker.jar", "sha-2");
        var artifacts = List.of(first, second);

        assertEquals(
                before.findByChecksum(List.of(), "missing"),
                after.findByChecksum(List.of(), "missing"));
        assertEquals(
                before.findByChecksum(artifacts, "missing"),
                after.findByChecksum(artifacts, "missing"));
        assertSame(
                first,
                after.findByChecksum(artifacts, "sha-1").orElseThrow());
        assertSame(
                second,
                after.findByChecksum(artifacts, "sha-2").orElseThrow());
        assertEquals(
                before.findByChecksum(artifacts, "sha-2"),
                after.findByChecksum(artifacts, "sha-2"));
    }

    @Test
    void bothVersionsReturnTheFirstMatchingInstance() {
        var first = new Artifact("api.jar", "same");
        var second = new Artifact("worker.jar", "same");
        var artifacts = List.of(first, second);

        assertSame(
                first,
                before.findByChecksum(artifacts, "same").orElseThrow());
        assertSame(
                first,
                after.findByChecksum(artifacts, "same").orElseThrow());
    }

    @Test
    void nullElementIsReadBeforeAMatchButNotAfterIt() {
        var matching = new Artifact("api.jar", "sha-1");
        List<Artifact> nullBeforeMatch = Arrays.asList(null, matching);
        List<Artifact> nullAfterMatch = Arrays.asList(matching, null);

        assertEquals(
                assertThrows(
                        NullPointerException.class,
                        () -> before.findByChecksum(nullBeforeMatch, "sha-1"))
                        .getMessage(),
                assertThrows(
                        NullPointerException.class,
                        () -> after.findByChecksum(nullBeforeMatch, "sha-1"))
                        .getMessage());
        assertSame(
                matching,
                before.findByChecksum(nullAfterMatch, "sha-1").orElseThrow());
        assertSame(
                matching,
                after.findByChecksum(nullAfterMatch, "sha-1").orElseThrow());
    }

    @Test
    void inputValidationHasTheSameOrderAndMessages() {
        assertSameFailure(null, "sha-1");
        assertSameFailure(List.of(), null);
    }

    @Test
    void earlyReturnPreservesIndexedListAccesses() {
        List<Artifact> artifacts = List.of(
                new Artifact("api.jar", "sha-1"),
                new Artifact("worker.jar", "sha-2"));
        List<String> legacyTrace = new ArrayList<>();
        List<String> refactoredTrace = new ArrayList<>();

        before.findByChecksum(traced(artifacts, legacyTrace), "sha-1");
        after.findByChecksum(traced(artifacts, refactoredTrace), "sha-1");

        assertEquals(List.of("size", "get:0"), legacyTrace);
        assertEquals(legacyTrace, refactoredTrace);
    }

    private void assertSameFailure(List<Artifact> artifacts, String checksum) {
        assertEquals(
                assertThrows(
                        NullPointerException.class,
                        () -> before.findByChecksum(artifacts, checksum))
                        .getMessage(),
                assertThrows(
                        NullPointerException.class,
                        () -> after.findByChecksum(artifacts, checksum))
                        .getMessage());
    }

    private static List<Artifact> traced(
            List<Artifact> artifacts,
            List<String> trace) {
        return new AbstractList<>() {
            @Override
            public Artifact get(int index) {
                trace.add("get:" + index);
                return artifacts.get(index);
            }

            @Override
            public int size() {
                trace.add("size");
                return artifacts.size();
            }
        };
    }
}

File: csharp/tests/Training.Module7.Tests/ReturnAsapEquivalenceTest.cs
using System.Collections;
using Training.Module7.ReturnAsap;
using Training.Module7.ReturnAsap.After;
using Training.Module7.ReturnAsap.Before;

namespace Training.Module7.Tests;

public sealed class ReturnAsapEquivalenceTest
{
    private readonly LegacyArtifactFinder _before = new();
    private readonly ArtifactFinder _after = new();

    [Fact]
    public void EarlyReturnPreservesEmptyMissingAndPresentResults()
    {
        var first = new Artifact("api.jar", "sha-1");
        var second = new Artifact("worker.jar", "sha-2");
        Artifact[] artifacts = [first, second];

        Assert.Equal(
            _before.FindByChecksum([], "missing"),
            _after.FindByChecksum([], "missing"));
        Assert.Equal(
            _before.FindByChecksum(artifacts, "missing"),
            _after.FindByChecksum(artifacts, "missing"));
        Assert.Same(first, _after.FindByChecksum(artifacts, "sha-1"));
        Assert.Same(second, _after.FindByChecksum(artifacts, "sha-2"));
        Assert.Equal(
            _before.FindByChecksum(artifacts, "sha-2"),
            _after.FindByChecksum(artifacts, "sha-2"));
    }

    [Fact]
    public void BothVersionsReturnTheFirstMatchingInstance()
    {
        var first = new Artifact("api.jar", "same");
        var second = new Artifact("worker.jar", "same");
        Artifact[] artifacts = [first, second];

        Assert.Same(first, _before.FindByChecksum(artifacts, "same"));
        Assert.Same(first, _after.FindByChecksum(artifacts, "same"));
    }

    [Fact]
    public void NullElementIsReadBeforeAMatchButNotAfterIt()
    {
        var matching = new Artifact("api.jar", "sha-1");
        Artifact[] nullBeforeMatch = [null!, matching];
        Artifact[] nullAfterMatch = [matching, null!];

        Assert.Equal(
            Assert.Throws<ArgumentNullException>(
                () => _before.FindByChecksum(nullBeforeMatch, "sha-1")).Message,
            Assert.Throws<ArgumentNullException>(
                () => _after.FindByChecksum(nullBeforeMatch, "sha-1")).Message);
        Assert.Same(matching, _before.FindByChecksum(nullAfterMatch, "sha-1"));
        Assert.Same(matching, _after.FindByChecksum(nullAfterMatch, "sha-1"));
    }

    [Fact]
    public void InputValidationHasTheSameOrderAndMessages()
    {
        AssertSameFailure(null, "sha-1");
        AssertSameFailure([], null);
    }

    [Fact]
    public void EarlyReturnPreservesIndexedListAccesses()
    {
        Artifact[] artifacts =
        [
            new("api.jar", "sha-1"),
            new("worker.jar", "sha-2")
        ];
        var legacyTrace = new List<string>();
        var refactoredTrace = new List<string>();

        _before.FindByChecksum(new TracedList(artifacts, legacyTrace), "sha-1");
        _after.FindByChecksum(new TracedList(artifacts, refactoredTrace), "sha-1");

        Assert.Equal(["size", "get:0"], legacyTrace);
        Assert.Equal(legacyTrace, refactoredTrace);
    }

    private void AssertSameFailure(IReadOnlyList<Artifact>? artifacts, string? checksum)
    {
        Assert.Equal(
            Assert.Throws<ArgumentNullException>(
                () => _before.FindByChecksum(artifacts!, checksum!)).Message,
            Assert.Throws<ArgumentNullException>(
                () => _after.FindByChecksum(artifacts!, checksum!)).Message);
    }

    private sealed class TracedList(IReadOnlyList<Artifact> artifacts, List<string> trace)
        : IReadOnlyList<Artifact>
    {
        public Artifact this[int index]
        {
            get
            {
                trace.Add($"get:{index}");
                return artifacts[index];
            }
        }

        public int Count
        {
            get
            {
                trace.Add("size");
                return artifacts.Count;
            }
        }

        public IEnumerator<Artifact> GetEnumerator() => artifacts.GetEnumerator();

        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
    }
}
